### PR TITLE
Added some of the features mentioned in the Types Widget Project.

### LINKED ISSUE #1078: Add listing typedefs in the panel ttj

### LINKED ISSUE #1079: Add listing structures in the panel tsj

### LINKED ISSUE #1080: Add listing union in the panel tuj

### LINKED ISSUE #1081: Add listing enum in the panel tej

### LINKED ISSUE #1082: Add filter capability in the panel

### PR DESCRIPTION
**Pull request by m0nk3ydluffy**
Added some of the features mentioned in the Types Widget Project.
Closes #1078 , closes #1079 , closes #1080 , closes #1081 , closes #1082 
To check the features:

1. You will need to use the console and use the "`to <path>`" command to open the file containing the types. 

2. Then you need to Refresh Contents from View Menu

Then you can see the custom types that you added from the file in the Types Widget

Will also be adding the feature to add the types from the GUI itself in the future.

Demo:
typeswidget

### CODE REVIEW

**Comment by ITAYC0HEN** [comment]
Thank you so much! will go over the code and test it 
Did you think of adding the `to <path>` command as a feature? So one can safely load a header file to Cutter

**Comment by m0nk3ydluffy** [comment]
Hey!
I will surely work on the feature to add the header file from the GUI itself - either by opening the file or by typing the descriptions into a textbox as said in issue #1074 .

**Comment by m0nk3ydluffy** [comment]
Hey @ITAYC0HEN . Thank you for reviewing. That was done very quickly. Will make the changes you requested.

**Comment by ITAYC0HEN** [comment]
Looks good to me! :) Elegant and helpful
Thank you so much, you are a real Pirate King ;)

File: src/widgets/TypesWidget.cpp [hunk 1]
```diff
@@ -32,9 +32,11 @@ QVariant TypesModel::data(const QModelIndex &index, int role) const
         case TYPE:
             return exp.type;
         case SIZE:
-            return exp.size;
+            return exp.category == "Primitive" ? exp.size : QVariant();
```
Comment (xarkes) [inline src/widgets/TypesWidget.cpp]: This won't work because you used `tr("Primitive")` earlier no?

File: src/widgets/TypesWidget.cpp [hunk 2]
```diff
@@ -75,7 +79,12 @@ bool TypesSortFilterProxyModel::filterAcceptsRow(int row, const QModelIndex &par
 {
     QModelIndex index = sourceModel()->index(row, 0, parent);
     TypeDescription exp = index.data(TypesModel::TypeDescriptionRole).value<TypeDescription>();
-    return exp.type.contains(filterRegExp());
+    if (selectedCategory.isEmpty()) {
+        return exp.type.contains(filterRegExp());
+    }
+    else {
```
Comment (xarkes) [inline src/widgets/TypesWidget.cpp]: No need to go to the newline? You can write `} else {` I think this is valid according to our coding style :)

File: src/widgets/TypesWidget.cpp (hunk 1 above)
Comment (m0nk3ydluffy) [reply src/widgets/TypesWidget.cpp]: Yep! Will correct this.

File: src/widgets/TypesWidget.cpp (hunk 2 above)
Comment (m0nk3ydluffy) [reply src/widgets/TypesWidget.cpp]: Okay! Cool

**Comment by XVilka** [comment]
What about adding context menu for adding a new type, which will open new textedit window to type the C type description (which you can feed to `td` command)? And second item in the menu can open a file open dialog for `to` command.

**Comment by ITAYC0HEN** [comment]
As far as I understood, @m0nk3ydluffy is planning to implement it next. And this is already tracked in issues

**Comment by m0nk3ydluffy** [comment]
Hey @XVilka ! As @ITAYC0HEN said, I will surely work on that next. There might be a delay of five to six days before I start working on due to my university exams.

**Pull request merged**